### ISSUE TITLE
Limit depth of recursion.

### ISSUE THREAD

**Issue by duesee** [duesee/imap-proto-server#3]

**Comment by duesee** [comment]
IMA4REV1 has three recursive parsers: `body`, `search-key`, and `sequence-set`.

```abnf
sequence-set = (seq-number / seq-range) ["," sequence-set]
```
is already parsed as

```Rust
separated_nonempty_list(tag(b","), alt((seq_range, seq_number)))
```
and does not recurse.

`search-key` has three variants, which recurse: `not`, `or`, and "and" (by using `(` and `)`).

Limiting the input does not work, as `(((((((((((` can easily trigger the recursion. Thus, we should transform the parser into an iterative one and introduce a `max_depth` parameter. 32 or even 16 should vastly suffice.

`body` is work in progress...

![Bildschirmfoto von 2020-08-27 10-03-05](https://user-images.githubusercontent.com/8997731/91421156-dfa1ab80-e855-11ea-9de7-0dac812366c2.png)

**Comment by duesee** [comment]
Fixed `search-key` in 098094198747fba94b6e3d4152a620cddfd7c031.

Because `body` is not implemented yet, I will remove the `bug` label for now, but keep the issue open.

**Comment by duesee** [comment]
Implemented `body` in https://github.com/duesee/imap-proto-server/commit/6f7c20253ee61cb82a69a11511d8aa136d59ec0a and introduced recursion limit in 1e15bdc5285bd20263ef79b47952af3161a00df2.
